File: OmniScript/cs/OmniScript/FileInformation.cs
// =============================================================================
// <copyright file="Adapter.cs" company="LiveAction, Inc.">
//  Copyright (c) 2018-2021 Savvius, Inc. All rights reserved.
//  Copyright (c) 2015-2018 Savvius, Inc. All rights reserved.
// </copyright>
// =============================================================================

namespace Savvius.Omni.OmniScript
{
    using System;
    using System.Runtime.InteropServices;
    using System.Collections.Generic;
    using System.IO;
    using System.Net;
    using System.Net.NetworkInformation;
    using System.Xml.Linq;
    using System.Text;

    public class FileInformation
    {
        public static readonly uint FlagsFolder = 0x00000010;

        public const int FILE_FLAGS_READONLY = 0x00000001;
        public const int FILE_FLAGS_HIDDEN = 0x00000002;
        public const int FILE_FLAGS_SYSTEM = 0x00000004;
        public const int FILE_FLAGS_FOLDER = 0x00000010;
        public const int FILE_FLAGS_ARCHIVE = 0x00000020;
        public const int FILE_FLAGS_DEVICE = 0x00000040;
        public const int FILE_FLAGS_NORMAL = 0x00000080;
        public const int FILE_FLAGS_TEMPORARY = 0x00000100;
        public const int FILE_FLAGS_SPARSE_FILE = 0x00000200;
        public const int FILE_FLAGS_REPARSE_POINT = 0x00000400;
        public const int FILE_FLAGS_COMPRESSED = 0x00000800;
        public const int FILE_FLAGS_OFFLINE = 0x00001000;
        public const int FILE_FLAGS_NOT_CONTENT_INDEXED = 0x00002000;
        public const int FILE_FLAGS_ENCRYPTED = 0x00004000;
        public const int FILE_FLAGS_VIRTUAL = 0x00010000;

        public OmniLogger Logger;

        public uint Flags;
        public PeekTime Modified;
        public String Path;
        public String Name;
        public UInt64 Size;

        // XML Tags
        const String tagRootName = "file";
        const String tagFlags = "attr";
        const String tagModified = "mod";
        const String tagName = "name";
        const String tagSize = "size";

        public FileInformation(OmniLogger logger, XElement node=null)
        {
            this.Logger = logger;
            this.Load(node);
        }

        public void Load(XElement node)
        {
            if (node == null) return;
            XElement info = null;
            if (node.Name.ToString() == tagRootName)
            {
                info = node;
            }
            if (info == null) return;

            IEnumerable<XAttribute> attribs = info.Attributes();
            foreach (XAttribute attrib in attribs)
            {
                String name = attrib.Name.ToString();
                switch (name)
                {
                    case tagFlags:
                        this.Flags = Convert.ToUInt32(attrib.Value);
                        break;

                    case tagModified:
                        this.Modified = new PeekTime(attrib.Value);
                        break;

                    case tagName:
                        this.Name = attrib.Value;
                        break;

                    case tagSize:
                        this.Flags = Convert.ToUInt32(attrib.Value);
                        break;
                }
            }
        }

        public XElement Store()
        {
            XElement file = new XElement(tagRootName,
                new XAttribute(tagName, this.Name),
                new XAttribute(tagFlags, this.Flags.ToString()),
                new XAttribute(tagSize, this.Size.ToString()));
            if (this.Modified != null)
            {
                file.Add(new XAttribute(tagModified, this.Modified.ToString()));
            }

            return file;
        }

        public void Store(XElement node)
        {
            node.Add(this.Store());
        }

        public bool IsFolder()
        {
            return ((this.Flags & FlagsFolder) != 0);
        }
    }
}

File: OmniScript/py/OmniScript/omniscript/fileinformation.py
"""FileInformation class.
"""
#Copyright (c) WildPackets, Inc. 2013-2014. All rights reserved.

import os
import sys
import logging
import xml.etree.ElementTree as ET

import omniscript
import config
from peektime import PeekTime


class FileInformation(object):
    """Information about a file.
    """

    flags = 0
    """The attribute flags of the file may be checked against the
    FILE FLAGS constants.
    """

    modified = None
    """The date and time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    that the file was last modified."""

    name = ''
    """The fully qualified name of the file."""

    size = 0
    """The size of the file in bytes."""

    #XML Tags
    _tag_root_name = "file"
    _tag_flags = "attr"
    _tag_modified = "mod"
    _tag_name = "name"
    _tag_size = "size"

    def __init__(self, name='', flags=0, elem=None):
        self.flags = flags
        self.modified = FileInformation.modified
        self.name = name
        self.size = FileInformation.size
        if elem is not None:
            self._load(elem)

    def _load(self, elem):
        """Load the File Informaiont from an ETree SubElement."""
        if elem.tag == FileInformation._tag_root_name:
            info = elem
        else:
            info = elem.find(FileInformation._tag_root_name)
        if info is not None:
            self.flags = int(info.attrib[FileInformation._tag_flags])
            self.modified = PeekTime(
                info.attrib[FileInformation._tag_modified])
            self.name = info.attrib[FileInformation._tag_name]
            self.size = int(info.attrib[FileInformation._tag_size])

    def __repr__(self):
        return 'FileInformation: %s' % self.name

    def is_folder(self):
        """Returns True if the information is for a folder."""
        return (self.flags & config.FILE_FLAGS_FOLDER) != 0

def file_information_list_to_file_list(fil):
    """Transform a list of
    :class:`FileInformation <omniscript.fileinformation.FileInformation>`
    into a list of strings of fully qualified file names.
    """
    fl = []
    path = ''
    for f in fil:
        if f.is_folder():
            path = f.name
        else:
            fl.append(path + f.name)
    return fl
